Add new scenario for camparison between Basic vs. Cbap Only schedulers

tommyazz opened the pull request.

tommyazz committed:
Add evaluation scenario for scheduling algorithms

tommyazz committed:
Add evaluation scenario for scheduling algorithms

tommyazz committed:
Add evaluation scenario for scheduling algorithms

tommyazz committed:
Add evaluation scenario for scheduling algorithms

tommyazz committed:
Add DmgTspec creation and function to compute SP duration

tommyazz committed:
Add DmgTspec creation and function to compute SP duration

tommyazz committed:
Add DmgTspec creation and function to compute SP duration

tommyazz committed:
Add DmgTspec creation and function to compute SP duration

tommyazz committed:
Add DmgTspec creation and function to compute SP duration

tommyazz committed:
Clean type definition for Assoc and DeAssoc callbacks at DmgStaWifiMac

tommyazz committed:
Clean type definition for Assoc and DeAssoc callbacks at DmgStaWifiMac

tommyazz committed:
Clean type definition for Assoc and DeAssoc callbacks at DmgStaWifiMac

tommyazz committed:
Clean type definition for Assoc and DeAssoc callbacks at DmgStaWifiMac

tommyazz committed:
Clean evaluation script and add MAC layer stats collection

tommyazz committed:
Clean evaluation script and add MAC layer stats collection

tommyazz committed:
Clean evaluation script and add MAC layer stats collection

tommyazz committed:
Clean evaluation script and add MAC layer stats collection

tommyazz committed:
Clean evaluation script and add MAC layer stats collection

tommyazz committed:
Small indentation issue

tommyazz committed:
Small indentation issue

tommyazz committed:
Small indentation issue

tommyazz committed:
Small indentation issue

tommyazz committed:
Revert "Modify ReportRxOk function and macRxOk callback "

This reverts commit 538763b0b3832f006d0be740f7b3d148d0b7c4ba.

tommyazz committed:
Modify ReportRxOk function and macRxOk callback 

Modify ReportRxOk function and macRxOk callback by adding a pointer to the successfully received packet as parameter.

tommyazz committed:
Modify ReportRxOk function and macRxOk callback 

Modify ReportRxOk function and macRxOk callback by adding a pointer to the successfully received packet as parameter.

tommyazz committed:
Modify ReportRxOk function and macRxOk callback 

Modify ReportRxOk function and macRxOk callback by adding a pointer to the successfully received packet as parameter.

tommyazz committed:
Revert "Modify ReportRxOk function and macRxOk callback "

This reverts commit 538763b0b3832f006d0be740f7b3d148d0b7c4ba.

tommyazz committed:
Revert "Modify ReportRxOk function and macRxOk callback "

This reverts commit 538763b0b3832f006d0be740f7b3d148d0b7c4ba.

tommyazz committed:
Modify ReportRxOk function and macRxOk callback 

Modify ReportRxOk function and macRxOk callback by adding a pointer to the successfully received packet as parameter.

tommyazz committed:
Modify ReportRxOk function and macRxOk callback 

Modify ReportRxOk function and macRxOk callback by adding a pointer to the successfully received packet as parameter.

tommyazz committed:
Revert "Modify ReportRxOk function and macRxOk callback "

This reverts commit 538763b0b3832f006d0be740f7b3d148d0b7c4ba.

tommyazz committed:
Enable SeqTsHeader at OnOffApplication and PackerSink

tommyazz committed:
Enable SeqTsHeader at OnOffApplication and PackerSink

tommyazz committed:
Enable SeqTsHeader at OnOffApplication and PackerSink

tommyazz committed:
Enable SeqTsHeader at OnOffApplication and PackerSink

tommyazz committed:
Clean Packet Sink code

if we receive a packet containing a SeqTs header we don't need to check for a Timestamp Tag since this information in included in the SeqTs header.

tommyazz committed:
Clean Packet Sink code

if we receive a packet containing a SeqTs header we don't need to check for a Timestamp Tag since this information in included in the SeqTs header.

tommyazz committed:
Clean Packet Sink code

if we receive a packet containing a SeqTs header we don't need to check for a Timestamp Tag since this information in included in the SeqTs header.

tommyazz committed:
Clean Packet Sink code

if we receive a packet containing a SeqTs header we don't need to check for a Timestamp Tag since this information in included in the SeqTs header.

tommyazz committed:
Add traced callback at Mac High for packets forwarded Up

Add timestamp tag to packets queued at Mac High

tommyazz committed:
Add traced callback at Mac High for packets forwarded Up

Add timestamp tag to packets queued at Mac High

tommyazz committed:
Revert "Add traced callback at Mac High for packets forwarded Up"

This reverts commit 144b2382d70af8d3d73da61066d21a0fc265b7bb.

tommyazz committed:
Revert "Add traced callback at Mac High for packets forwarded Up"

This reverts commit 144b2382d70af8d3d73da61066d21a0fc265b7bb.

tommyazz committed:
Add traced callback at Mac High for packets forwarded Up

Add timestamp tag to packets queued at Mac High

tommyazz committed:
Add traced callback at Mac High for packets forwarded Up

Add timestamp tag to packets queued at Mac High

tommyazz committed:
Revert "Add traced callback at Mac High for packets forwarded Up"

This reverts commit 144b2382d70af8d3d73da61066d21a0fc265b7bb.

tommyazz committed:
Add traced callback at Mac High for packets forwarded Up

Add timestamp tag to packets queued at Mac High

tommyazz committed:
Add traced callback at Mac High for packets forwarded Up

Add timestamp tag to packets queued at Mac High

tommyazz committed:
Add code measure APP and MAC layer throughput and delay

Fix some minor issues. Default app=onoff, default socket=udp

tommyazz committed:
Add code measure APP and MAC layer throughput and delay

Fix some minor issues. Default app=onoff, default socket=udp

tommyazz committed:
Add code measure APP and MAC layer throughput and delay

Fix some minor issues. Default app=onoff, default socket=udp

tommyazz committed:
Add code measure APP and MAC layer throughput and delay

Fix some minor issues. Default app=onoff, default socket=udp

tommyazz committed:
Adjust MacRxOk callback. Add interallocation distance as tunable param

tommyazz committed:
Adjust MacRxOk callback. Add interallocation distance as tunable param

tommyazz committed:
Adjust MacRxOk callback. Add interallocation distance as tunable param

tommyazz committed:
Adjust MacRxOk callback. Add interallocation distance as tunable param

tommyazz committed:
Fix issue in DmgWifiHelper

 The dmg scheduler factory needs to be created by default otherwise when the scheduler object is never created.

Minor fixes in qd dense scenario

tommyazz committed:
Fix issue in DmgWifiHelper

 The dmg scheduler factory needs to be created by default otherwise when the scheduler object is never created.

Minor fixes in qd dense scenario

tommyazz committed:
Fix issue in DmgWifiHelper

 The dmg scheduler factory needs to be created by default otherwise when the scheduler object is never created.

Minor fixes in qd dense scenario

tommyazz committed:
Fix issue in DmgWifiHelper

 The dmg scheduler factory needs to be created by default otherwise when the scheduler object is never created.

Minor fixes in qd dense scenario

tommyazz committed:
Remove timestamp Tag before a packet is queued at Mac High

tommyazz committed:
Remove timestamp Tag before a packet is queued at Mac High

tommyazz committed:
Remove timestamp Tag before a packet is queued at Mac High

tommyazz committed:
Remove timestamp Tag before a packet is queued at Mac High

tommyazz committed:
Remove timestamp Tag before a packet is queued at Mac High

tommyazz committed:
Remove timestamp Tag before a packet is queued at Mac High

tommyazz committed:
Add SeqTs header to Bulk Application

tommyazz committed:
Add SeqTs header to Bulk Application

tommyazz committed:
Add SeqTs header to Bulk Application

tommyazz committed:
Add SeqTs header to Bulk Application

tommyazz committed:
Fix issue: non-associated STAs were able to  participate in an SP

Non-associated STAs (with AID=0 before association) were able to start an SP as destination or source in those SPs where the intended source or destination is the PCP/AP (AID=0). Now, only associated can participate in an SP as source, destination or relay.

tommyazz committed:
Fix issue: non-associated STAs were able to  participate in an SP

Non-associated STAs (with AID=0 before association) were able to start an SP as destination or source in those SPs where the intended source or destination is the PCP/AP (AID=0). Now, only associated can participate in an SP as source, destination or relay.

tommyazz committed:
Fix issue: non-associated STAs were able to  participate in an SP

Non-associated STAs (with AID=0 before association) were able to start an SP as destination or source in those SPs where the intended source or destination is the PCP/AP (AID=0). Now, only associated can participate in an SP as source, destination or relay.

tommyazz committed:
Fix issue: non-associated STAs were able to  participate in an SP

Non-associated STAs (with AID=0 before association) were able to start an SP as destination or source in those SPs where the intended source or destination is the PCP/AP (AID=0). Now, only associated can participate in an SP as source, destination or relay.

tommyazz committed:
Add isSource parameter to service period callback

tommyazz committed:
Add isSource parameter to service period callback

tommyazz committed:
Add isSource parameter to service period callback

tommyazz committed:
Add isSource parameter to service period callback

tommyazz committed:
Small changes to evaluate scheduler scenario

tommyazz committed:
Small changes to evaluate scheduler scenario

tommyazz committed:
Small changes to evaluate scheduler scenario

tommyazz committed:
Small changes to evaluate scheduler scenario

mattia-lecci committed:
Add new codebooks

Adding new codebooks generated with MATLAB.
Tapering (kaiser window) has been used to obtain these specific configurations and, in general, to obtain codebook with a taget size.
Codewords have been normalized to unit norm.

mattia-lecci committed:
Add new codebooks

Adding new codebooks generated with MATLAB.
Tapering (kaiser window) has been used to obtain these specific configurations and, in general, to obtain codebook with a taget size.
Codewords have been normalized to unit norm.

mattia-lecci committed:
Add new codebooks

Adding new codebooks generated with MATLAB.
Tapering (kaiser window) has been used to obtain these specific configurations and, in general, to obtain codebook with a taget size.
Codewords have been normalized to unit norm.

mattia-lecci committed:
Add new codebooks

Adding new codebooks generated with MATLAB.
Tapering (kaiser window) has been used to obtain these specific configurations and, in general, to obtain codebook with a taget size.
Codewords have been normalized to unit norm.

tommyazz committed:
Add Jitter computation at Packet Sink

tommyazz committed:
Add Jitter computation at Packet Sink

tommyazz committed:
Add Jitter computation at Packet Sink

tommyazz committed:
Add Jitter computation at Packet Sink

tommyazz committed:
Add Jitter. Create output file for the simulation. Clean code

tommyazz committed:
Add Jitter. Create output file for the simulation. Clean code

tommyazz committed:
Add Jitter. Create output file for the simulation. Clean code

tommyazz committed:
Add Jitter. Create output file for the simulation. Clean code

tommyazz committed:
Comment cmd params for SEM. Add first campaign script

tommyazz committed:
Comment cmd params for SEM. Add first campaign script

tommyazz committed:
Comment cmd params for SEM. Add first campaign script

tommyazz committed:
Comment cmd params for SEM. Add first campaign script

tommyazz committed:
Add sem script. Add absolute path to load DmgFiles folder

tommyazz committed:
Add sem script. Add absolute path to load DmgFiles folder

tommyazz committed:
Add sem script. Add absolute path to load DmgFiles folder

tommyazz committed:
Add sem script. Add absolute path to load DmgFiles folder

tommyazz committed:
Merge remote-tracking branch 'origin/feature/scheduler-evaluation-scenario' into feature/scheduler-evaluation-scenario

tommyazz committed:
Remove beamforming in Cbap in the evaluation scenario. Update sem script

tommyazz committed:
Remove beamforming in Cbap in the evaluation scenario. Update sem script

tommyazz committed:
Remove beamforming in Cbap in the evaluation scenario. Update sem script

tommyazz committed:
Removed Snr output stream

tommyazz committed:
Removed Snr output stream

tommyazz committed:
Removed Snr output stream

tommyazz committed:
Modified Sem script

tommyazz committed:
Modified Sem script

tommyazz committed:
Modified Sem script

tommyazz committed:
v2 campaign

tommyazz committed:
v2 campaign

tommyazz committed:
v2 campaign

MatteoDrago commented on src/wifi/helper/dmg-wifi-helper.cc:
```diff
@@ -169,6 +169,7 @@ DmgWifiHelper::DmgWifiHelper ()
 {
   SetRemoteStationManager ("ns3::ConstantRateWifiManager", "ControlMode", StringValue ("DMG_MCS4"),
                                                            "DataMode", StringValue ("DMG_MCS12"));
+  SetDmgScheduler ("ns3::CbapOnlyDmgWifiScheduler");
```
so we choose CBAP only by default?

mattia-lecci replied:
at least for the moment i think it's fine, at least it's basically the same as when there was no scheduler at all

tommyazz replied:
Seems legit

MatteoDrago commented on src/wifi/model/dmg-sta-wifi-mac.cc:
```diff
@@ -1593,7 +1592,7 @@ DmgStaWifiMac::StartFullDuplexRelay (AllocationID allocationID, Time length,
   m_peerStationAddress = peerAddress;
   m_moreData = true;
-  m_servicePeriodStartedCallback (GetAddress (), peerAddress);
+  m_servicePeriodStartedCallback (GetAddress (), peerAddress, isSource);
```
does this mean that this method is called twice? once here, and once in DmgWifiMac::StartServicePeriod? Is this a bug or a feature?

tommyazz replied:
Let me investigate. I don't know how the relay procedure works

tommyazz replied:
So, `m_servicePeriodStartedCallback`, here, signals the start of a Full Duplex SP during relay operations, which is unrelated to the "classical" workflow of operations (i.e. without relaying). 
If Relay operations are not supported, in `DmgStaWifiMac::InitiateAllocationPeriod` the method `DmgWifiMac::StartServicePeriod` is called, and the callback in question is triggered one time.

MatteoDrago commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+ * ./waf --run "evaluate_scheduler_qd_dense_scenario"
+ *
+ * Simulation Output:
+ * The simulation generates the following traces:
+ *
+ */
```
Is this incomplete?

MatteoDrago commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+
+void
+DataTransmissionIntervalStarted (Ptr<DmgStaWifiMac> wifiMac, Mac48Address address, Time)
+{
+  /* Avoiding beamforming training in a CBAP for the moment */
+  /*if (wifiMac->IsAssociated ())
+    {
+      uint16_t counter = biCounter[address];
+      counter++;
+      if (counter == biThreshold)
+        {
+          wifiMac->InitiateTxssCbap (wifiMac->GetBssid ());
+          counter = 0;
+        }
+      biCounter[address] = counter;
+    }
+    */
+}
```
Delete if not needed

MatteoDrago commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+  tcpVariants.insert (make_pair ("Westwood",      "ns3::TcpWestwood"));
+  tcpVariants.insert (make_pair ("WestwoodPlus",  "ns3::TcpWestwoodPlus"));
+
+  /* Command line argument parser setup. */
+  CommandLine cmd;
+  // cmd.AddValue ("applicationType", "Type of the Tx Application: onoff or bulk", applicationType);
+  // cmd.AddValue ("packetSize", "Application packet size [bytes]", packetSize);
+  cmd.AddValue ("dataRate", "Application data rate", appDataRate);
+  // cmd.AddValue ("maxPackets", "Maximum number of packets to send", maxPackets);
+  // cmd.AddValue ("tcpVariant", "Transport protocol to use: TcpHighSpeed, TcpVegas, TcpNewReno, TcpWestwood, TcpWestwoodPlus", tcpVariant);
+  // cmd.AddValue ("socketType", "Socket type (default: ns3::UdpSocketFactory)", socketType);
+  // cmd.AddValue ("bufferSize", "TCP Buffer Size (Send/Receive) [bytes]", bufferSize);
+  // cmd.AddValue ("msduAggregation", "The maximum aggregation size for A-MSDU [bytes]", msduAggregationSize);
+  cmd.AddValue ("mpduAggregation", "The maximum aggregation size for A-MPDU [bytes]", mpduAggregationSize);
+  // cmd.AddValue ("queueSize", "The maximum size of the Wifi MAC Queue [packets]", queueSize);
+  // cmd.AddValue ("frameCapture", "Use a frame capture model", frameCapture);
+  // cmd.AddValue ("frameCaptureMargin", "Frame capture model margin [dB]", frameCaptureMargin);
+  cmd.AddValue ("phyMode", "802.11ad PHY Mode", phyMode);
+  // cmd.AddValue ("verbose", "turn on all WifiNetDevice log components", verbose);
+  cmd.AddValue ("simulationTime", "Simulation time [s]", simulationTime);
+  // cmd.AddValue ("qdChannelFolder", "The name of the folder containing the QD-Channel files", qdChannelFolder);
+  // cmd.AddValue ("numSTAs", "The number of DMG STA", numSTAs);
+  // cmd.AddValue ("pcap", "Enable PCAP Tracing", pcapTracing);
+  cmd.AddValue ("scheduler", "The type of scheduler to use in the simulation", schedulerType);
+  cmd.AddValue ("interAllocation", "Duration of a broadcast CBAP between two ADDTS allocations [us]", interAllocDistance);
+  // cmd.AddValue ("csv", "Enable CSV output instead of plain text. This mode will suppress all the messages related statistics and events.", csv);
+  cmd.AddValue ("logComponentsStr", "Components to be logged from tLogStart to tLogEnd separated by ':'", logComponentsStr);
+  // cmd.AddValue ("tLogStart", "Log start [s]", tLogStart);
+  // cmd.AddValue ("tLogEnd", "Log end [s]", tLogEnd);
+  cmd.Parse (argc, argv);
```
What about cleaning up all the unused cmd? Are you keeping them "just in case"?

tommyazz replied:
Yes!
I commented most of them for compatibility with SEM. We can restore them maybe

MatteoDrago commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+        }
+      cout << "\nAggregate Throughput: " << aggregateThr << endl;  
+
+      /* Print MAC layer Stats */
+      /* cout << "\nMAC Layer Statistics:" << endl;
+      for (PacketCountMap::const_iterator it = macTxDataOk.begin (); it != macTxDataOk.end (); ++it)
+        {
+          NS_LOG_UNCOND ("DMG Station: " << it->first);
+          NS_LOG_UNCOND ("Tx Packets: " << it->second);
+          NS_LOG_UNCOND ("Tx Failed : " << macTxDataFailed.at (it->first));
+          NS_LOG_UNCOND ("Rx Packets: " << macRxDataOk.at (it->first));
+        }
+      */
```
Delete if no needed

MatteoDrago commented on ad-first-campaign.py:
```diff
@@ -0,0 +1,66 @@
+def main():
```
as discussed, let's put it into another branch :)

tommyazz replied:
Yes

mattia-lecci commented on src/wifi/model/wifi-remote-station-manager.h:
```diff
@@ -809,14 +809,15 @@ class WifiRemoteStationManager : public Object
    */
   void ReportRxOk (Mac48Address address, const WifiMacHeader *header,
-                   double rxSnr, WifiMode txMode);
+                   Ptr<const Packet> packet, double rxSnr, WifiMode txMode);
```
is this needed? it doesn't look so from the code but i might be wrong

mattia-lecci commented on src/wifi/model/wifi-remote-station-manager.h:
```diff
@@ -1606,7 +1607,7 @@ class WifiRemoteStationManager : public Object
   Callback<void, Mac48Address> m_rxCallbackOk;
 
-  TracedCallback<WifiMacType, Mac48Address, double> m_macRxOk;   /* Trace for Successful WifiMAC Reception */
+  TracedCallback<WifiMacType, Ptr<const Packet>, Mac48Address, double> m_macRxOk;   /* Trace for Successful WifiMAC Reception */
```
same comment as above

mattia-lecci commented on src/wifi/model/wifi-remote-station-manager.h:
```diff
@@ -1615,7 +1616,7 @@ class WifiRemoteStationManager : public Object
    * \param snr The snr value in linear scale.
    */
-  typedef void (* MacRxOkCallback)(WifiMacType type, Mac48Address address, double snr);
+  typedef void (* MacRxOkCallback)(WifiMacType type, Ptr<const Packet> packet, Mac48Address address, double snr);
```
same comment

mattia-lecci commented on src/wifi/model/wifi-remote-station-manager.cc:
```diff
@@ -1126,7 +1126,7 @@ WifiRemoteStationManager::ReportRxOk (Mac48Address address, const WifiMacHeader
   WifiRemoteStation *station = Lookup (address, header);
   m_rxCallbackOk (address);
-  m_macRxOk (header->GetType (), address, rxSnr);
+  m_macRxOk (header->GetType (), packet->Copy (), address, rxSnr);
```
why copy? packet is constant so you shouldn't be able to modify it anyway and its existence should be granted by ns3's smart pointers

mattia-lecci commented on src/wifi/model/regular-wifi-mac.h:
```diff
@@ -803,6 +803,10 @@ class RegularWifiMac : public WifiMac
 
+  /* Packet forwarded Up traced callback */
+  TracedCallback<Ptr<Packet>, Mac48Address, Mac48Address> m_forwardUpTrace;
+  typedef void (* ForwardUpTracedCallback)(Ptr<Packet> packet, Mac48Address from, Mac48Address to);
```
is this still needed? i think we said to avoid looking at metrics from lower layers, right?

mattia-lecci commented on src/wifi/model/dmg-wifi-mac.h:
```diff
@@ -870,13 +870,14 @@ class DmgWifiMac : public RegularWifiMac
    * \param srcAddress The MAC address of the source station.
    * \param dstAddress The MAC address of the destination station.
    */
```
missing doc param

mattia-lecci commented on src/wifi/model/dmg-wifi-mac.h:
```diff
@@ -870,13 +870,14 @@ class DmgWifiMac : public RegularWifiMac
-  typedef void (* ServicePeriodCallback)(Mac48Address srcAddress, Mac48Address dstAddress);
-  TracedCallback<Mac48Address, Mac48Address> m_servicePeriodStartedCallback;
-  TracedCallback<Mac48Address, Mac48Address> m_servicePeriodEndedCallback;
+  typedef void (* ServicePeriodCallback)(Mac48Address srcAddress, Mac48Address dstAddress, bool isSource);
```
is this needed? you already have src and sdt addresses

tommyazz replied:
It can be helpful for easy debugging. What do you think? :)

mattia-lecci replied:
i'm not sure, how would you intend to use it? maybe i'm missing something

tommyazz replied:
It's just that the `ServicePeriodCallback` can be fired also when an SP starts at the destination node (of the SP). It might be useful when debugging with many STAs. It's not essential. If we don't want to modify too much existing modules let's avoid this.

mattia-lecci replied:
no I think it's fine, it's not the Wifi core module, if you think it's useful then let's keep it

tommyazz committed:
Address Matteo's feedback. Clean up code

tommyazz committed:
Address Matteo's feedback. Clean up code

tommyazz committed:
Address Matteo's feedback. Clean up code

mattia-lecci commented on src/applications/model/packet-sink.cc:
```diff
@@ -66,19 +66,22 @@ PacketSink::GetTypeId (void)
 }
 
 PacketSink::PacketSink ()
```
damn this code was a mess :)

mattia-lecci commented on src/applications/model/packet-sink.cc:
```diff
@@ -66,19 +66,22 @@ PacketSink::GetTypeId (void)
+    m_totalRx (0),
+    m_totalPackets (0),
+    m_delayAccumulator (Seconds (0)),
+    m_jitterAccumulator (Seconds (0))   
```
missing `m_currentDelay`

mattia-lecci commented on src/applications/model/packet-sink.cc:
```diff
@@ -202,20 +208,33 @@ void PacketSink::HandleRead (Ptr<Socket> socket)
+          if (m_currentDelay != 0)
+            {
+              /* Accumulate Jitter value */
+              m_jitterAccumulator += Seconds (std::abs (m_currentDelay.GetSeconds () - delay.GetSeconds ()));
```
this is more like "delay accumulation" rather than "jitter"

mattia-lecci commented on src/applications/model/packet-sink.h:
```diff
@@ -138,7 +144,9 @@ class PacketSink : public Application
   TracedCallback<Ptr<const Packet>, const Address &> m_rxTrace;
-  Time accummulator;
+  Time m_delayAccumulator;       //!< Accumulate the delay of the currently received packet
+  Time m_jitterAccumulator;      //!< Accumulate the jitter if the currently received packet
```
i'm really not a fan of having the jitter hard-coded into the `PacketSink` class, nor of modifying core-classes. Can't we just collect the delay traces and compute it outside ns-3?

mattia-lecci replied:
also, have you checked the class `DelayJitterEstimationTimestampTag`?
I'm not sure if it can be of any use but i would avoid re-writing code

mattia-lecci commented on src/applications/model/packet-sink.cc:
```diff
@@ -202,20 +208,33 @@ void PacketSink::HandleRead (Ptr<Socket> socket)
-      // Should never not be found since the sender is adding it, but
-      // you never know.
-      if (packet->FindFirstMatchingByteTag (timestamp))
+      /* If we have a SeqTs header we don't need to check for a Timestamp tag  */
```
is the timestamp tag going to be there anyway? if so, there's no need to look at the `SeqTsHeader `and thus to modify the code

mattia-lecci commented on src/applications/model/packet-sink.cc:
```diff
@@ -202,20 +208,33 @@ void PacketSink::HandleRead (Ptr<Socket> socket)
+        {
+          /* Otherwise, look for the presence of a Timestamp tag */
+          TimestampTag timestamp;
+          if (packet->FindFirstMatchingByteTag (timestamp))
```
no jitter computation if app is using timestamp? i think tag and seqTs header should be treated the same..

mattia-lecci commented on src/applications/model/onoff-application.cc:
```diff
@@ -301,16 +301,16 @@ void OnOffApplication::SendPacket ()
+  NS_ABORT_IF (packet->GetSize () != m_pktSize);
+  NS_LOG_DEBUG ("Tx seq=" << header.GetSeq () << ", at time=" << header.GetTs ());
+
+// Ptr<Packet> packet = Create<Packet> (m_pktSize);
```
i see something similar has been done in ns3-dev so ok

mattia-lecci replied:
but actually, as discussed offline, probably a tag would be better. seqts header should only be used for udp really, and anyway you're adding it at the app layer.. it looks a bit fishy to me

mattia-lecci commented on src/applications/model/bulk-send-application.cc:
```diff
@@ -193,8 +194,15 @@ void BulkSendApplication::SendData (void)
+      packet->AddHeader (header);
+      NS_ABORT_IF (packet->GetSize () != toSend);
+      NS_LOG_DEBUG ("Tx seq=" << header.GetSeq () << ", at time=" << header.GetTs ());
+
```
same as for onoff

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+double
+CalculateSingleStreamThroughput (Ptr<PacketSink> sink, uint64_t &lastTotalRx, double &averageThroughput)
+{
+  double thr = (sink->GetTotalRx () - lastTotalRx) * (double) 8 / 1e5;     /* Convert Application RX Packets to Mbits. */
```
Also, why 1e5?
```suggestion
  double thr = (sink->GetTotalRx () - lastTotalRx) *  8.0 / 1e5;     /* Convert Application RX Packets to Mbits. */
```

tommyazz replied:
It should be because `(sink->GetTotalRx () - lastTotalRx) *  8.0` is bits/100ms. Then times 10 is bits/s and divided by 10^6 is Mbits/s.

mattia-lecci replied:
oh jesus, please fix this mess :D
it can easily lead to stupid bugs and it is not clear at all looking at the code.
I see that the 100ms period is hard coded. You could set it as a global variable and then use it to compute the `thr` with the appropriate EXPLICIT transformations

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+};
+typedef map<Ptr<Node>, CommunicationPair> CommunicationPairList;
+typedef CommunicationPairList::iterator CommunicationPairListI;
+typedef CommunicationPairList::const_iterator CommunicationPairListCI;
```
you could easily use `auto`, it's actually a best practice :)

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+{
+  double thr = (sink->GetTotalRx () - lastTotalRx) * (double) 8 / 1e5;     /* Convert Application RX Packets to Mbits. */
+  lastTotalRx = sink->GetTotalRx ();
+  averageThroughput += thr;
```
not sure what you're doing: this is actually totRxMbits, not an avg throughput

tommyazz replied:
Sorry Mattia, this comes from Hany's code and I've never checked it. Fortunately, this code just displays the throughput every 100 ms in the standard output. Any issue won't affect the results we collected for the NIST's report. According to my comment above this should thr in Mbps. Let me know :)

mattia-lecci replied:
ok so the throughput over 100 ms windows is being added.. doesn't look like an average..
it is saved into `CommunicationPair.throughput` so I guess this is not what he wants to do.
I'd say either of these would be ok:
1. rename variables in a more meaningful way
2. remove that field from `CommunicationPair` since it is not used (but it should be done for all scratch scripts..)
3. compute and the average throughput
An average can be updated online if you know how many elements you have been averaging.
Starting from i=0, T_{0,avg}=0, you can compute T_{n,avg} = (n-1)/n * T_{n-1,avg} + T_n/n, for each n>1. In this case you could either save and update `i` as a global variable or compute it as `i= Simulator::Now () / MilliSeconds(100)`

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+        {
+          node = it->second.srcApp->GetNode ();
+          thr = CalculateSingleStreamThroughput (it->second.packetSink, it->second.totalRx, it->second.throughput);
+          totalThr += thr;
```
again, this is actually totRxMbits since the previous call, not a throughput

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+NS_LOG_COMPONENT_DEFINE ("EvaluateScheduler");
+
+using namespace ns3;
+using namespace std;
```
please avoid

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+    {
+      string duration = to_string_with_precision<double> (Simulator::Now ().GetSeconds () - 0.1, 1) +
+                        " - " + to_string_with_precision<double> (Simulator::Now ().GetSeconds (), 1);
+      cout << left << setw (12) << duration;
```
`NS_LOG_UNCOND` should be used instead

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+        {
+          NS_FATAL_ERROR ("Could not find application to start.");
+        }
+    }
```
at some point we should add this code into `DmgStaWifiMac` :)
maybe add a TODO just to remember about it

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+{
+  NS_LOG_FUNCTION (dataBitRate << phyModeDataRate);
+  double numberBIs = Seconds (1).GetMicroSeconds () / double (apWifiMac->GetBeaconInterval ().GetMicroSeconds ());
+  uint32_t spDuration = ceil (dataBitRate / double (phyModeDataRate) / numberBIs * 1e6);
```
```suggestion
  uint64_t dataRateRatio = dataBitRate / phyModeDataRate;
  uint64_t biDuration_us = apWifiMac->GetBeaconInterval ().GetMicroSeconds ();
  uint32_t spDuration = dataRateRatio * biDuration_us;
```

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+  NS_LOG_FUNCTION (dataBitRate << phyModeDataRate);
+  double numberBIs = Seconds (1).GetMicroSeconds () / double (apWifiMac->GetBeaconInterval ().GetMicroSeconds ());
+  uint32_t spDuration = ceil (dataBitRate / double (phyModeDataRate) / numberBIs * 1e6);
+  return spDuration + 1500;
```
why adding 1.5ms? just to be sure?

tommyazz replied:
Yes! This was a guess. And it just holds for this scenario

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+  if (it != communicationPairList.end ())
+    {
+      uint32_t spDuration = ComputeServicePeriodDuration (it->second.appDataRate, WifiMode (phyMode).GetPhyRate ());
+      staWifiMac->CreateAllocation (GetDmgTspecElement (allocationId++, true, spDuration, spDuration));
```
doesn't every STA have its own allocation ID? Why are you incrementing it?

tommyazz replied:
No, the allocationID is a parameter associated to the DMG Tspec. Each STA has its own AID.

mattia-lecci replied:
yeah, but i mean: a TS is identified by the tuple (srcAid, dstAid, allocationId), right?
Since you have N STAs with a single TS each, you could keep a constant allocationId.
Think about what would happen if there were more than 16 STAs: allocationId would overflow (4 bits if i remember correctly)

tommyazz replied:
Oh sure! That's right

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+}
+
+void
+StationAssociated (Ptr<Node> node, Ptr<DmgStaWifiMac> staWifiMac, Mac48Address address, uint16_t aid)
```
```suggestion
StationAssociated (Ptr<Node> node, Ptr<DmgStaWifiMac> staWifiMac, Mac48Address apAddress, uint16_t aid)
```

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+}
+
+void
+StationDeAssociated (Ptr<Node> node, Ptr<DmgWifiMac> staWifiMac, Mac48Address address)
```
```suggestion
StationDeAssociated (Ptr<Node> node, Ptr<DmgWifiMac> staWifiMac, Mac48Address apAddress)
```

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+}
+
+CommunicationPair
+InstallApplications (Ptr<Node> srcNode, Ptr<Node> dstNode, Ipv4Address address, string appDataRate, uint16_t appNumber)
```
```suggestion
InstallApplication (Ptr<Node> srcNode, Ptr<Node> dstNode, Ipv4Address address, string appDataRate, uint16_t appNumber)
```

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+      srcApp = src.Install (srcNode);
+    }
+  /* The application is started when the corresponding ADDTS request succeeded (or failed only for CbapOnlyDmgWifiScheduler) */
+  srcApp.Start (Seconds (simulationTime + 1));
```
is this needed?

tommyazz replied:
Yes, otherwise the app will start at 0 by default.

mattia-lecci replied:
ok, maybe improve the comment then, it is not clear enough IMO

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+}
+
+void
+DataTransmissionIntervalStarted (Ptr<DmgStaWifiMac> wifiMac, Mac48Address address, Time)
```
remove if not needed

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+  double frameCaptureMargin = 10;                 /* Frame capture margin [dB]. */
+  bool verbose = false;                           /* Print Logging Information. */
+  bool pcapTracing = false;                       /* Enable PCAP Tracing. */
+  uint16_t numSTAs = 8;                          /* The number of DMG STAs. */
```
there are 10 STAs in the picture above (in the doc). Also, please write the max number of STAs supported

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+  EnableMyTraces (logComponents, Seconds (tLogStart), Seconds (tLogEnd));
+  LogComponentEnable ("EvaluateScheduler", LOG_LEVEL_ALL);
+
+  string currentPath = SystemPath::FindSelfDirectory ();
+  vector<string> vectorPath = SplitString (currentPath, '/');
+  string inputPath = "/";
+  string dir;
+  for (uint32_t i = 0; i < vectorPath.size (); ++i)
+    {
+      dir = vectorPath[i];
+      if (dir == "")
+        continue;
+      inputPath += dir + "/";
+      if (dir == "ns3-802.11ad")
+        break;
+    }
+  NS_LOG_UNCOND (inputPath);
```
create a function

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+  /* Print Output */
+  if (!csv)
+    {
+      cout << "Application Layer Throughput per Communicating Pair [Mbps]" << endl;
```
`NS_LOG_UNCOND`

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+uint32_t msduAggregationSize = 7935;          /* The maximum aggregation size for A-MSDU [bytes]. */
+uint32_t mpduAggregationSize = 262143;        /* The maximum aggregation size for A-MPDU [bytes]. */
+double simulationTime = 10;                   /* Simulation time [s]. */
+bool csv = false;                             /* Enable CSV output. */
```
i would remove this. looks like it's a way to print to `std::cout` in csv format and remove some logging. csv cout makes no sense (use traces instead), logging is automatically removed when running simulations is optimized build profile if `NS_LOG_UNCOND` is used in place of `std::cout`

mattia-lecci commented on ad-first-campaign.py:
```diff
@@ -0,0 +1,66 @@
+    ############################
+
+    script = "evaluate_scheduler_qd_dense_scenario"
+    cmp_name = "basic-vs-cbaponly-v2"        # define a campaign name in order to create different corresponding to a specific configuration
```
not that much longer but definitely more clear :)
```suggestion
    campaign_name = "basic-vs-cbaponly-v2"        # define a campaign name in order to create different corresponding to a specific configuration
```

mattia-lecci commented on ad-first-campaign.py:
```diff
@@ -0,0 +1,66 @@
+
+    script = "evaluate_scheduler_qd_dense_scenario"
+    cmp_name = "basic-vs-cbaponly-v2"        # define a campaign name in order to create different corresponding to a specific configuration
+    ns_path = os.path.join(os.path.dirname(os.path.realpath(__file__)))
```
is `os.path.join` needed?

mattia-lecci commented on ad-first-campaign.py:
```diff
@@ -0,0 +1,66 @@
+    import sem
+    import math
+    import numpy as np
+    import os
```
maybe
```suggestion
    from os import path
```
will suffice

mattia-lecci commented on ad-first-campaign.py:
```diff
@@ -0,0 +1,66 @@
+    script = "evaluate_scheduler_qd_dense_scenario"
+    cmp_name = "basic-vs-cbaponly-v2"        # define a campaign name in order to create different corresponding to a specific configuration
+    ns_path = os.path.join(os.path.dirname(os.path.realpath(__file__)))
+    campaign_dir = "./campaigns/"+cmp_name
```
be consistent with the style
```suggestion
    campaign_dir = "./campaigns/" + cmp_name
```

mattia-lecci commented on ad-first-campaign.py:
```diff
@@ -0,0 +1,66 @@
+    campaign = sem.CampaignManager.new(ns_path, script, campaign_dir,
+        runner_type="ParallelRunner",
+        check_repo=False,
+        skip_config=True,
```
is this in sem's official release?

mattia-lecci commented on ad-first-campaign.py:
```diff
@@ -0,0 +1,66 @@
+    dataRate = "100Mbps"
+    mcs = "DMG_MCS4"
+    simulationTime = 10.0
+    logComponentsStr = ":"
```
you could comment this parameter as well in the cc script

mattia-lecci commented on ad-first-campaign.py:
```diff
@@ -0,0 +1,66 @@
+    campaign.run_missing_simulations (param_combinations)
+
+    print("Simulations done.")
+
+if __name__ == '__main__':
+    main()
```
not needed

mattia-lecci commented on ad-first-campaign.py:
```diff
@@ -0,0 +1,66 @@
+def main():
```
not needed if treating it like a main script (see other comment)

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -0,0 +1,825 @@
+/*
```
i only checked this scratch script, the same comments apply for all other scripts

mattia-lecci reviewed (commented):
good job, but the core has been modified a bit too much IMO

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -249,19 +229,19 @@ ADDTSResponseReceived (Ptr<Node> node, Mac48Address address, StatusCode status,
   NS_LOG_FUNCTION (+allocId << isPseudoStatic << minAllocation << maxAllocation);
+  /* Simple assert for the moment */
   NS_ASSERT_MSG (minAllocation <= maxAllocation, "Minimum Allocation cannot be greater than Maximum Allocation");
   NS_ASSERT_MSG (maxAllocation <= MAX_SP_BLOCK_DURATION, "Maximum Allocation exceeds Max SP block duration");
```
assert are removed in optimized mode. you might want to use abort if instead

mattia-lecci reviewed (commented):
damn i didn't see this commit :(

tommyazz committed:
Remove python simulation script

tommyazz committed:
Remove python simulation script

tommyazz committed:
Remove python simulation script

tommyazz committed:
Revert "Enable SeqTsHeader at OnOffApplication and PackerSink"

This reverts commit b34fc46bb85e9c62903b8f8dbb46624ed1baf1b9.

tommyazz committed:
Revert "Enable SeqTsHeader at OnOffApplication and PackerSink"

This reverts commit b34fc46bb85e9c62903b8f8dbb46624ed1baf1b9.

tommyazz committed:
Revert "Enable SeqTsHeader at OnOffApplication and PackerSink"

This reverts commit b34fc46bb85e9c62903b8f8dbb46624ed1baf1b9.

tommyazz committed:
Revert "Add SeqTs header to Bulk Application"

This reverts commit aac8b7fb836fb274fa42337b66cb70ba3ab82308.

tommyazz committed:
Revert "Add SeqTs header to Bulk Application"

This reverts commit aac8b7fb836fb274fa42337b66cb70ba3ab82308.

tommyazz committed:
Revert "Add SeqTs header to Bulk Application"

This reverts commit aac8b7fb836fb274fa42337b66cb70ba3ab82308.

tommyazz committed:
Address Mattia's feeback. Restore packet sink to its original code

tommyazz committed:
Address Mattia's feeback. Restore packet sink to its original code

tommyazz committed:
Address Mattia's feeback. Restore packet sink to its original code

tommyazz committed:
Restore cmd parameters

tommyazz committed:
Restore cmd parameters

tommyazz committed:
Restore cmd parameters

tommyazz committed:
Clean code of L-Room evaluation scenario. Minor fixes Scheduler dense

tommyazz committed:
Clean code of L-Room evaluation scenario. Minor fixes Scheduler dense

tommyazz committed:
Clean code of L-Room evaluation scenario. Minor fixes Scheduler dense

tommyazz committed:
Remove accidentally pushed results file

tommyazz committed:
Remove accidentally pushed results file

tommyazz committed:
Remove accidentally pushed results file

tommyazz committed:
Move unnecessary scripts from scratch to examples/wireless/80211ad

tommyazz committed:
Move unnecessary scripts from scratch to examples/wireless/80211ad

tommyazz committed:
Move unnecessary scripts from scratch to examples/wireless/80211ad

mattia-lecci commented on ad-first-campaign.py:
```diff
@@ -1,66 +0,0 @@
-def main():
```
i don't remember if i said to remove it, and in any case why? It's never going to be merge into master anyway

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -96,53 +89,53 @@ struct CommunicationPair
+uint32_t mpduAggregationSize = 262143;             /* The maximum aggregation size for A-MPDU [bytes]. */
+double simulationTime = 10;                        /* Simulation time [s]. */
+uint8_t allocationId = 1;                          /* The allocation ID of the DMG Tspec element to create */
+uint16_t thrLogPeriodicity = 100;                  /* The log periodicity for the throughput of each STA [ms] */
```
would improve the readability in some parts of the code
```suggestion
Time thrLogPeriodicity = MilliSeconds (100);                  /* The log periodicity for the throughput of each STA [ms] */
```

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -166,12 +159,43 @@ EnableMyTraces (vector<string> &logComponents, Time tLogStart, Time tLogEnd)
 {
-  double thr = (sink->GetTotalRx () - lastTotalRx) * (double) 8 / 1e5;     /* Convert Application RX Packets to Mbits. */
+  double rxBits = (sink->GetTotalRx () - lastTotalRx) * 8.0; /* Total Rx Bits in the last period */
+  double rxBitsPerSec = rxBits * (1e3 / thrLogPeriodicity);  /* Total Rx bits per second */
```
see previous comment

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -180,20 +204,19 @@ CalculateThroughput (void)
 {
   double totalThr = 0;
   double thr;
+  std::string duration = to_string_with_precision<double> (Simulator::Now ().GetSeconds () - (double (thrLogPeriodicity) / 1e3), 2) +
```
see previous comment

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -180,20 +204,19 @@ CalculateThroughput (void)
-  cout << left << setw (12) << duration;
-
-  for (CommunicationPairListI it = communicationPairList.begin (); it != communicationPairList.end (); ++it)
+  for (auto it = communicationPairList.begin (); it != communicationPairList.end (); ++it)
```
you could even use a foreach, but it's not necessary :)

mattia-lecci commented on scratch/evaluate_scheduler_qd_dense_scenario.cc:
```diff
@@ -232,18 +256,21 @@ uint32_t
+  uint64_t biDurationUs = apWifiMac->GetBeaconInterval ().GetMicroSeconds ();
+  uint32_t spDuration = ceil (dataRateRatio * biDurationUs);
+
+  return spDuration;
```
was the extra 1500 removed on purpose?

tommyazz replied:
Yes, it was specific for the simulation campaign we conducted

mattia-lecci reviewed (commented):
it's basically ready for me, just change the target branch

tommyazz committed:
Minor fixes and improvements

tommyazz committed:
Minor fixes and improvements

tommyazz committed:
Minor fixes and improvements

tommyazz committed:
Merge remote-tracking branch 'origin/feature/scheduler-evaluation-scenario' into feature/scheduler-evaluation-scenario

tommyazz committed:
Add support to period SP requests in evaluation scripts

tommyazz committed:
Add support to period SP requests in evaluation scripts

tommyazz committed:
Minor fixes. Add received packets trace

tommyazz committed:
Add useful trace for debugging

tommyazz committed:
Fix issue with new output traces

tommyazz committed:
Printing Time in current raw resolution

tommyazz committed:
Add output stream for AC_BE queue's size at MAC

tommyazz committed:
Add contention period started/ended traced callbacks

tommyazz committed:
Add contention period started/ended traced callbacks

mattia-lecci committed:
Using 1.3 factor for spDuration

Adding a constant term equal to 1500us is not a good practice. We noticed that the APP rate is about 21% less than the PHY rate, so we increase the SP duration

tommyazz committed:
Resolving rebase issues
